src/couchdb/rolling-database.ts

```diff
@@ -0,0 +1,393 @@
+      }
+    } = opts
+
+    // Ensure we have a list database:
+    const listDbSetup: DatabaseSetup = {
+      name: `${name}-list`,
+      replicatorSetup,
+      onChange() {
+        readDbList().catch(onError)
+      }
+    }
+    const listDbCleanup = await setupDatabase(connection, listDbSetup, opts)
+    const listDb = connection.use(listDbSetup.name)
+
+    /**
+     * Reads and processes the list database's contents.
+     */
+    const readDbList = withMutex(async function readDbList(): Promise<void> {
+      const now = new Date()
+
+      const response = await listDb.list({ include_docs: true })
+      let list: RollingDatabaseList = response.rows
+        .map(row => {
+          const clean = asListEntry(row.doc)
+          return {
+            archived: clean.doc.archived,
+            name: clean.id,
+            startDate: clean.doc.startDate
+          }
+        })
+        .sort((a, b) => b.startDate.valueOf() - a.startDate.valueOf())
+
+      async function addDb(
+        name: string,
+        startDate: Date,
+        archived = false
+      ): Promise<void> {
+        await listDb.insert(
+          wasListEntry({ id: name, doc: { archived, startDate } })
+        )
+        list = [{ archived, name, startDate }, ...list]
+      }
+
+      // Create past databases if the list is empty:
+      if (list.length === 0) {
+        if (archiveStart == null) {
+          const [date, suffix] = pickPeriodicMonth(now, period, false)
+          await addDb(`${name}-${suffix}`, date)
+        } else {
+          for (
+            let [date, suffix] = pickPeriodicMonth(archiveStart, 'year', false);
+            date.valueOf() < now.valueOf();
+            [date, suffix] = pickPeriodicMonth(date, 'year', true)
+          ) {
+            await addDb(`${name}-${suffix}`, date)
+          }
+        }
+      }
+
+      // Always keep one spare database in the future:
+      const lastStart = list[0].startDate
+      if (lastStart.valueOf() < now.valueOf()) {
+        const [date, suffix] = pickPeriodicMonth(now, period, true)
+        await addDb(`${name}-${suffix}`, date)
+      }
+
+      // Ensure that each new database exists:
+      cleanups.forEach(cleanup => cleanup())
+      cleanups = []
+      for (const { archived, name } of list) {
+        const setup: DatabaseSetup = { name, ...setupRest }
+        if (!archived) setup.replicatorSetup = replicatorSetup
+        cleanups.push(await setupDatabase(connection, setup, opts))
+      }
+
+      // Make the list available to queries:
+      databases = list
+    })
+
+    // Do the initial processing:
+    await readDbList()
+
+    // Periodically create new databases:
+    const creationTask = makePeriodicTask(
+      async () => await readDbList(),
+      24 * 60 * 60 * 1000,
+      { onError }
+    )
+    if (!disableWatching) creationTask.start({ wait: true })
+
+    // Return a cleanup function:
+    return () => {
+      cleanups.forEach(cleanup => cleanup())
```

Comment:
Do I understand correctly: `cleanups` is an array of scheduled database transactions? Took me a bit to understand this because of the name; unless I still don't understand the semantics here.

Reply:
The `cleanups` array handles the same sort of stuff `componentWillUnmount` would normally do. Basically, we want to close our web sockets, cancel our timers, and de-allocate other resources so the app can shut down gracefully. Nothing in the `cleanups` array visits the server, aside from closing any open connections.

Reply:
Ah, I see. I misread line 358 as pushing the setupDatabase routine in the array, not the cleaner it returns.